File: Problems/No.108_Convert_Sorted_Array_to_Binary_Search_Tree/convert_sorted_array_to_binary_search_tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    
    // make binary search balanced tree
    TreeNode* makeBST(vector<int>& nums, int left_index, int right_index) {
        
        if (left_index > right_index) return NULL;
        else if (left_index == right_index) return new TreeNode(nums[left_index]);
        else {
            
            int mid = (left_index+right_index)/2;
            TreeNode *left = makeBST(nums, left_index, mid - 1);
            TreeNode *right = makeBST(nums, mid + 1, right_index);
            return new TreeNode(nums[mid], left, right);
            
        }
        
    }
    
    TreeNode* sortedArrayToBST(vector<int>& nums) {
        
        return makeBST(nums, 0, nums.size()-1);
        
    }
};

File: Problems/No.108_Convert_Sorted_Array_to_Binary_Search_Tree/convert_sorted_array_to_binary_search_tree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def sortedArrayToBST(self, nums):
        """
        :type nums: List[int]
        :rtype: TreeNode
        """
        
        if nums:
            
            mid = len(nums)//2
            new_node = TreeNode(nums[mid])
            new_node.left = self.sortedArrayToBST(nums[:mid])
            new_node.right = self.sortedArrayToBST(nums[mid+1:])
            
            return new_node
        
